src/types/tags.ts

```diff
@@ -13,6 +13,20 @@ export type QueryTag<
   key: QueryTagKey
 }
 
+export type QueryTagType<TQueryTag extends QueryTag> = TQueryTag extends QueryTag<infer TData> 
+  ? TData extends Unset
+    ? unknown
+    : TData
+   : never
+
+export function isQueryTag(tag: unknown): tag is QueryTag {
+  return typeof tag === 'object' && tag !== null && 'data' in tag && 'key' in tag
+}
```

Comment:
this seems mostly safe, but given pretty generic property names it seems plausible it will yield false positives. Thoughts on adding some sort of internal symbol, like we've done in the past to mark tags

Reply:
Yeah we might want to do something like that. For now this is fine since it's just used to distinguish between an action and a tag